### PR TITLE
Updates amp-analytics with support for expanding config/trigger vars.

### PR DESCRIPTION
**Pull request by btownsend**
Adds support for expanding place holders in amp-analytics request template strings with values specified in amp-analytics trigger vars and config vars. This also changes the place holder syntax  to ${id} (instead of {id}) for consistency with ES6 template strings.

### CODE REVIEW

**Comment by btownsend** [comment]
CLA should already be signed.

File: src/string.js [hunk 1]
```diff
@@ -24,3 +24,33 @@ export function dashToCamelCase(name) {
     return character.toUpperCase();
   });
 }
+
+/**
+ * Expands place holders in a given template string with values.
```
Comment (dvoytenko) [inline src/string.js]: "placeholders"?

File: src/string.js [hunk 2]
```diff
@@ -24,3 +24,33 @@ export function dashToCamelCase(name) {
...
+ *
+ * Place holders use ${key-name} syntax and are replaced with the value
+ * returned from the given getter function.
+ *
+ * @param {string} template The template string to expand.
+ * @param {!function(string):*} getter Function used to retrieve a value for a
+ *   place holder. Returns values will be coerced into strings.
+ * @param {number=1} optMaxIterations Number of times to expand the template.
```
Comment (dvoytenko) [inline src/string.js]: opt_maxIterations

File: examples/analytics.amp.html [hunk 3]
```diff
@@ -2,78 +2,45 @@
...
-
-  Curabitur mattis libero a venenatis porta. Integer ac aliquam nulla. Nunc non orci sit amet metus consectetur fermentum. Suspendisse et velit sit amet nulla rhoncus semper. Vivamus et dolor tempus, ultricies metus at, lacinia nisi. Nunc justo dui, efficitur ut orci eget, vehicula feugiat est. Suspendisse non urna congue, feugiat est nec, tempus nisi. In convallis vitae lectus eu finibus. Vestibulum ullamcorper id turpis id pulvinar. Nulla facilisi. Aenean purus odio, imperdiet sed nisl ut, luctus ullamcorper leo. Aenean sit amet pharetra nunc. Integer maximus diam vitae tortor lobortis, sed facilisis ante vehicula. Mauris rutrum pharetra pretium. Praesent laoreet faucibus odio, quis fringilla libero convallis quis.
+<amp-analytics id="analytics1">
+{
+  "host": "example.com",
+  "requests": {
+    "default": "/log?domain=DOMAIN&amp;path=PATH"
```
Comment (dvoytenko) [inline examples/analytics.amp.html]: Will you add any examples with `${x}` form?

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 4]
```diff
@@ -176,37 +177,13 @@ export class AmpAnalytics extends AMP.BaseElement {
     }
     this.requests_ = requests;
 
-    // Now replace any request templates with their values.
-    const all = Object.keys(this.requests_).join('|');
-    const requestExpr = new RegExp('\{(' + all + ')\}', 'g');
+    // Expand any placeholders. For requests, we expand each string up to 5
+    // times to support nested requests. Leave any unresolved place holders.
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Could you pls rename "place holder" -> "placeholder" everywhere?

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 5]
```diff
@@ -224,7 +201,17 @@ export class AmpAnalytics extends AMP.BaseElement {
       return;
     }
 
-    // TODO(btownsend, #871): Add support for variables from inline config.
+    // Replace place holders with URI encoded values.
+    // Precedence is trigger.vars > config.vars > (TODO) platform.vars.
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Please reference name and issue in a TODO

File: extensions/amp-analytics/0.1/vendors.js [hunk 6]
```diff
@@ -24,20 +24,27 @@ export const ANALYTICS_CONFIG = {
   'googleanalytics': {
     'host': 'www.google-analytics.com',
     'method': 'GET',
+    'vars': {
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/vendors.js]: How does this set reconcile against UrlReplacements? And why does it have some but not other vars from there?

**Comment by dvoytenko** [comment]
Thanks! I think this looks good. Just a few comments.

File: src/string.js (hunk 1 above)
Comment (btownsend) [inline src/string.js]: Done.

File: src/string.js (hunk 2 above)
Comment (btownsend) [inline src/string.js]: Done.

File: examples/analytics.amp.html (hunk 3 above)
Comment (btownsend) [inline examples/analytics.amp.html]: Done.

File: extensions/amp-analytics/0.1/amp-analytics.js (hunk 5 above)
Comment (btownsend) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Done.

File: extensions/amp-analytics/0.1/vendors.js (hunk 6 above)
Comment (btownsend) [inline extensions/amp-analytics/0.1/vendors.js]: This is temporary. Added a TODO explaining the long term plan.

File: extensions/amp-analytics/0.1/amp-analytics.js (hunk 4 above)
Comment (btownsend) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Done.

**Comment by btownsend** [comment]
Updated to address feedback.

**Comment by dvoytenko** [comment]
LGTM. Could you please squash commits?

**Comment by btownsend** [comment]
Squashed.

**Pull request merged**